# Мэтчинг товаров для маркетплейса

## Открытие файлов с данными и их изучение

Маркетплейс поставил перед нами задачу мэтчинга: необходимо разработать алгоритм, который для каждого товара из датасета `validation.csv` предложит 5 наиболее похожих товаров из `base.csv`. Оценивать качество алгоритма нужно оценивать по метрике accuracy@5 по ответам из датасета `validation_answer.csv`.  
  
Нам доступно 4 датасета с данными:
- `base.csv` - анонимизированный набор товаров. Каждый товар представлен как уникальный id и вектор признаков размерностью 72
- `train.csv` - датасет с данными для обучения модели. У каждого товара есть уникальный id, вектор признаков размерностью 72 и id товара из датасета `base.csv`, который максимально похож на него по мнению экспертов
- `validation.csv` - датасет с набором товаров, для которых надо найти наиболее близкие товары из `base.csv`. У каждого товара есть уникальный id и набор признаков размерностью 72
- `validation_answer.csv` - правильные ответы к предыдущему файлу

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV, PredefinedSplit, cross_val_score
!pip install faiss-gpu
import faiss
import requests
from zipfile import ZipFile
import urllib
import json
from urllib.parse import urlencode
import time
from google.colab import files
!pip  install catboost
from catboost import CatBoostClassifier, Pool
import os

In [2]:
base_url = 'https://cloud-api.yandex.net/v1/disk/public/resources/download?'
public_key = 'https://disk.yandex.ru/d/BBEphK0EHSJ5Jw'

# Формируем ссылку с яндекс диском для загрузки
final_url = base_url + urlencode(dict(public_key=public_key))
response = requests.get(final_url)
download_url = response.json()['href']

# Загружаем файл и сохраняем его
download_response = requests.get(download_url)
with open('data.zip', 'wb') as f:
    f.write(download_response.content)

# Сохраняем zip file, в котором хранятся данные
zip_file = ZipFile('data.zip')

In [3]:
train = pd.read_csv(zip_file.open('train.csv'), index_col=0) # Обучающий датасет
base = pd.read_csv(zip_file.open('base.csv'), index_col=0) # Набор товаров
validation = pd.read_csv(zip_file.open('validation.csv'), index_col=0) # Товары, для которых ищем наиболее похожие
answers = pd.read_csv(zip_file.open('validation_answer.csv'), index_col=0) # Ответы к предыдущему сету товаров

In [4]:
# Удалим ненужные файлы для очистки памяти
del zip_file, download_response

delete_filepath = '/content/data.zip'

open(delete_filepath, 'w').close()
os.remove(delete_filepath)

In [5]:
def data_overview(data, corr=False):

    # Настроим вывод датасетов, чтобы показывались все столбцы
    pd.set_option('display.max_columns', None)

    # Вывод первых строчек датасета для лучшего понимания структуры
    display(data.head())

    # Выведем размер датасетае
    print(f'Размер датасета: {data.shape}')

    # Проверим датасет на наличие пропусков
    nulls = pd.DataFrame(data.isna().sum(), columns=['NA_count'])
    if sum(nulls['NA_count'] == 0):
        print('В датасете нет пропусков')
    else:
        display(nulls.query('NA_count != 0'))

    # Выведем информацию о распределении признаков
    display(data.describe())

    # Посмотрим на корреляцию между признаками
    if corr:
        display(data.corr())

In [6]:
# Обучающий датасет 'train'
data_overview(train)

0          1           2          3           4           5  \
Id                                                                             
0-query -53.882748  17.971436  -42.117104 -183.93668  187.517490  -87.144930   
1-query -87.776370   6.806268  -32.054546 -177.26039  120.803330  -83.810590   
2-query -49.979565   3.841486 -116.118590 -180.40198  190.128430  -50.837620   
3-query -47.810562   9.086598 -115.401695 -121.01136   94.652840 -109.255410   
4-query -79.632126  14.442886  -58.903397 -147.05254   57.127068  -16.239529   

                  6          7           8           9         10          11  \
Id                                                                              
0-query -347.360606  38.307602  109.085560   30.413513  -88.08269  -52.695540   
1-query  -94.572749 -78.433090  124.915900  140.331070 -177.60580  -84.995514   
2-query   26.943937 -30.447489  125.771164  211.607820  -86.34656  -35.666546   
3-query -775.150134  79.186520  124.003100  242.650650 -146.51707 -159.469850   
4-query -321.317964  45.984676  125.941284  103.392670 -107.15302   -8.800034   

                12         13          14         15         16          17  \
Id                                                                            
0-query -27.692442   4.872923  198.348000 -30.075249  -3.803569 -135.810610   
1-query  42.810810 -57.256332   96.792534 -19.261467   0.739535   50.619213   
2-query  16.395317 -80.802850  137.908650 -23.532760 -47.256584  -16.650242   
3-query -13.844755  -6.113928  118.939255 -44.585907   9.559358   14.435648   
4-query -50.977800  29.457338  143.389310   5.614824 -45.274760    9.643625   

                18         19        20           21         22          23  \
Id                                                                            
0-query -161.84137 -78.512180 -94.42894   898.436927 -70.140520   78.420360   
1-query -155.26703 -78.659430 -92.76149   353.157741 -34.744545   82.487110   
2-query -194.50568 -78.372925 -69.32448  1507.231274 -52.500970  -34.165775   
3-query -156.90683 -78.789320 -78.73709  1507.231274  19.957405   34.834290   
4-query  -77.55463 -79.066610 -77.92646  1507.231274  16.612400  116.284290   

                 24          25          26         27          28         29  \
Id                                                                              
0-query  108.032776  813.770071  -74.790880  12.610422 -183.821840  149.71584   
1-query  -28.450592  813.770071 -137.529630  26.595627 -136.783450  153.35791   
2-query   52.958652  813.770071  -18.021725  20.951107  -50.321780  158.76062   
3-query   -8.820732  813.770071 -125.606800  17.584084  -58.452904  141.28180   
4-query   33.754898  813.770071 -105.765335   6.523008  -19.812988  157.69392   

                30         31          32           33        34         35  \
Id                                                                            
0-query -27.042316 -133.21217  106.420746  -303.939233  48.73079  58.185707   
1-query  48.810093 -115.92215   87.464220  -222.286354  25.12415  91.887140   
2-query   0.178065 -183.06967   99.053570 -1018.469545 -51.80112  97.766770   
3-query -54.959310 -136.98854   63.880493 -1018.469545  89.22893  65.919960   
4-query -20.604088 -146.59128   78.849570  -780.449185  87.56077  73.036660   

                36         37          38         39         40        41  \
Id                                                                          
0-query  24.250950 -151.22410 -104.282265 -34.492810 -12.587054  2.622891   
1-query -30.636870 -136.59314 -140.500120 -43.449757  -7.226884  8.265747   
2-query -10.865850 -144.42316 -133.819490 -78.902300 -17.200352  4.467452   
3-query -24.078644 -152.33410  -91.199380 -28.225390  -4.767386  0.158236   
4-query  16.891030 -144.65790 -116.122150 -19.353254  -7.709266 -5.394988   

                 42         43         44          45          46         47  \
Id                                    

Размер датасета: (100000, 73)
В датасете нет пропусков


0              1              2              3  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -85.328679       7.664345     -43.667046    -146.118630   
std        25.803845       4.955651      39.111064      20.434841   
min      -186.280270     -11.560507    -224.896060    -223.307220   
25%      -103.309147       4.313784     -69.494487    -160.028520   
50%       -85.273695       7.652854     -42.830246    -146.067445   
75%       -67.330810      10.981191     -17.713690    -132.277745   
max        14.585236      28.917845     128.108460     -60.751625   

                   4              5              6              7  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      111.770592     -73.178792    -440.615118      21.231588   
std        47.700958      28.718853     279.017550      65.985214   
min       -93.272020    -184.962450    -791.469482    -278.582820   
25%        79.474322     -92.834080    -738.536868     -22.131764   
50%       112.260100     -73.364420    -511.837758      22.278989   
75%       143.760830     -53.699612    -201.116578      65.899595   
max       301.363600      51.849710     109.625188     288.171840   

                   8              9             10             11  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      123.067891     126.120201    -141.168584     -69.228444   
std         6.492081      64.912752      41.607785      51.258647   
min        94.831150    -142.156950    -330.570600    -352.405200   
25%       118.674642      82.545028    -168.599020    -103.048815   
50%       123.087540     125.963340    -140.461440     -68.753325   
75%       127.541802     169.701005    -113.136535     -34.668559   
max       149.356200     392.827150      37.017212     152.500000   

                  12             13             14             15  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -42.559541       6.506719     149.077582     -23.942253   
std        40.819544      37.593800      34.167843      13.024184   
min      -214.742460    -145.860280       8.818390     -82.645004   
25%       -69.801705     -18.815189     126.369124     -32.379752   
50%       -41.919796       6.739488     149.705675     -23.707227   
75%       -14.631057      32.074534     172.627318     -15.438074   
max       116.210100     171.126970     279.152560      35.552612   

                  16             17             18             19  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean       -2.219093     -12.964439    -138.714610     -79.159795   
std        20.978943      55.210920      48.737245       0.775560   
min       -91.657020    -247.589280    -326.648740     -82.637184   
25%       -16.455603     -49.612492    -171.661355     -79.656816   
50%        -2.339008     -12.189709    -139.009150     -79.137460   
75%        11.702535      24.464662    -106.746862     -78.638890   
max        94.324580     215.488050      79.245160     -75.835640   

                  20             21             22             23  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -97.474574    1297.269578      -3.600685      55.908014   
std        28.382144     375.123794      50.063680      35.617513   
min      -228.808260     136.873137    -211.971800     -87.323780   
25%      -115.407545    1248.126198     -37.047511      31.436791   
50%       -96.862695    1507.231274      -3.608930      55.273705   
75%       -78.583615    1507.231274      29.473402      79.557540   
max        26.484070    1557.383334     193.884310     207.021060   

                  24             25             26             27  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean       23.470181     742.788817     -64.970553      16.368072   
std        40.802653     191.052074      63.160650       4.828575   
min      -1

In [7]:
# Набор товаров 'base'
data_overview(base)

0          1          2          3           4           5  \
Id                                                                            
0-base -115.083890  11.152912 -64.426760 -118.88089  216.482440 -104.698060   
1-base  -34.562202  13.332763 -69.787610 -166.53348   57.680607  -86.098370   
2-base  -54.233746   6.379371 -29.210136 -133.41383  150.895830  -99.435326   
3-base  -87.520130   4.037884 -87.803030 -185.06763   76.369540  -58.985165   
4-base  -72.743850   6.522049  43.671265 -140.60803    5.820023 -112.074080   

                 6          7           8           9         10          11  \
Id                                                                             
0-base -469.070588  44.348083  120.915344  181.449700 -124.06151  -32.458237   
1-base  -85.076666 -35.637436  119.718636  195.234190 -141.30435 -126.768900   
2-base   52.554795  62.381706  128.951450  164.381470 -140.82245  -22.486748   
3-base -383.182845 -33.611237  122.031910  136.233580 -108.47044 -104.537780   
4-base -397.711282  45.182500  122.167180  112.119064 -110.34245  -38.982060   

               12         13         14         15         16          17  \
Id                                                                          
0-base -57.420560  36.207405  182.71677 -28.071688  -5.209374   -1.732182   
1-base -68.853530   8.314717  143.08174   8.778257  15.157185   48.246360   
2-base -51.666990   2.521726  140.58545 -22.039230  -7.128634   -5.184787   
3-base -82.388490 -79.986740  113.59631 -38.995880  -4.787360    8.327808   
4-base -50.228867 -26.690920  135.27290  -7.510103   8.924209  111.576950   

               18        19          20           21          22         23  \
Id                                                                            
0-base  -91.09186 -79.46667 -103.939090  1507.231274  -32.591667  51.414360   
1-base -244.95486 -80.27942  -81.788040   429.880035   -5.937690  70.142654   
2-base -111.71212 -79.02927  -82.452576  1507.231274   -0.421011  84.108010   
3-base -124.85576 -79.10902 -107.943275  1479.170530 -125.271614  24.656485   
4-base -130.83319 -79.31585  -57.843952  1262.919738   19.225910  59.811813   

               24          25          26         27          28         29  \
Id                                                                            
0-base -51.905360  813.770071  -41.556538   8.419807 -129.976640  148.76503   
1-base  66.576840  813.770071    3.214371  17.164303  -39.418385  148.15588   
2-base  63.344510  813.770071 -105.211050  18.844700  -94.789474  151.38771   
3-base -16.921055  813.770071 -182.680270  14.715704  -62.932700  151.20198   
4-base  76.159670  312.698903  -76.726974  21.299923  -24.069107  167.38553   

               30         31          32          33         34          35  \
Id                                                                            
0-base -10.069234 -143.00504   79.848694 -537.183707   2.773180  111.518680   
1-base  -7.579590 -174.36995  101.224630 -303.939233  16.914630   53.394676   
2-base -21.483440 -144.84537   48.502934 -780.449185  64.767310   79.781555   
3-base  10.613454 -134.93434   58.777730 -529.295053  43.665924   92.806305   
4-base -44.059470 -148.13185   98.042610 -657.524388  96.363170   71.061520   

               36         37         38         39         40         41  \
Id                                                                         
0-base   4.257666 -153.34058 -123.62608 -39.707664 -12.768708  -4.170106   
1-base  33.451824 -148.55263 -110.34017 -33.277206  -8.688703  13.650302   
2-base -13.390142 -150.06166 -118.75855 -40.896286 -15.222620  -1.860338   
3-base -38.936657 -145.04530 -124.29577 -36.876730 -17.742870  10.536242   
4-base -36.563557 -135.37244 -123.70898 -44.739437 -10.855220   8.354800   

                42          43         44          45          46         47  \
Id                                                                             
0-base 

Размер датасета: (2918139, 72)
В датасете нет пропусков


0             1             2             3             4  \
count  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06   
mean  -8.622947e+01  8.080077e+00 -4.458080e+01 -1.466350e+02  1.113166e+02   
std    2.489132e+01  4.953387e+00  3.863166e+01  1.984480e+01  4.634809e+01   
min   -1.994687e+02 -1.391461e+01 -2.400734e+02 -2.326671e+02 -1.055830e+02   
25%   -1.030654e+02  4.708491e+00 -6.955949e+01 -1.599051e+02  8.050795e+01   
50%   -8.623150e+01  8.038950e+00 -4.381661e+01 -1.467768e+02  1.118730e+02   
75%   -6.925658e+01  1.147007e+01 -1.962527e+01 -1.333277e+02  1.423743e+02   
max    2.151555e+01  2.993721e+01  1.609372e+02 -5.137478e+01  3.196645e+02   

                  5             6             7             8             9  \
count  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06   
mean  -7.199138e+01 -3.922239e+02  2.035283e+01  1.236842e+02  1.244581e+02   
std    2.818607e+01  2.716550e+02  6.421638e+01  6.356109e+00  6.443058e+01   
min   -2.110086e+02 -7.914699e+02 -3.018597e+02  9.315305e+01 -1.738719e+02   
25%   -9.137994e+01 -6.293318e+02 -2.222147e+01  1.194840e+02  8.176751e+01   
50%   -7.192230e+01 -4.222016e+02  2.080477e+01  1.238923e+02  1.234977e+02   
75%   -5.244111e+01 -1.566686e+02  6.391821e+01  1.279705e+02  1.672206e+02   
max    5.880624e+01  1.096325e+02  3.412282e+02  1.522612e+02  4.275421e+02   

                 10            11            12            13            14  \
count  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06   
mean  -1.431396e+02 -6.988662e+01 -4.126493e+01  7.587780e+00  1.489852e+02   
std    4.168737e+01  5.121991e+01  3.896358e+01  3.659725e+01  3.363559e+01   
min   -3.438436e+02 -3.684526e+02 -2.260906e+02 -1.756239e+02 -1.526698e+01   
25%   -1.703175e+02 -1.040333e+02 -6.662606e+01 -1.694393e+01  1.259664e+02   
50%   -1.417156e+02 -7.022748e+01 -4.039738e+01  8.128445e+00  1.493146e+02   
75%   -1.149862e+02 -3.569530e+01 -1.483560e+01  3.267095e+01  1.721625e+02   
max    4.140846e+01  1.824389e+02  1.207674e+02  1.775270e+02  2.917480e+02   

                 15            16            17            18            19  \
count  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06   
mean  -2.316701e+01 -9.366916e-01 -1.931312e+01 -1.372712e+02 -7.918932e+01   
std    1.307563e+01  2.083201e+01  5.297210e+01  4.662566e+01  7.574403e-01   
min   -8.591711e+01 -9.636816e+01 -2.725656e+02 -3.517478e+02 -8.315235e+01   
25%   -3.176832e+01 -1.515450e+01 -5.546463e+01 -1.675851e+02 -7.967751e+01   
50%   -2.291030e+01 -1.032792e+00 -1.980470e+01 -1.368043e+02 -7.917803e+01   
75%   -1.434341e+01  1.337403e+01  1.658952e+01 -1.075373e+02 -7.867896e+01   
max    4.507333e+01  1.053910e+02  2.207994e+02  9.768367e+01 -7.579646e+01   

                 20            21            22            23            24  \
count  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06   
mean  -9.825383e+01  1.257528e+03 -6.824936e+00  5.667382e+01  2.669563e+01   
std    2.815861e+01  3.999612e+02  5.197251e+01  3.436061e+01  3.894157e+01   
min   -2.437279e+02  1.368182e+02 -2.382874e+02 -8.818329e+01 -1.431833e+02   
25%   -1.168426e+02  1.088597e+03 -4.224413e+01  3.329542e+01  5.564670e-01   
50%   -9.752848e+01  1.507231e+03 -6.418125e+00  5.699871e+01  2.713143e+01   
75%   -7.924342e+01  1.507231e+03  2.823144e+01  7.966341e+01  5.308961e+01   
max    3.212392e+01  1.557433e+03  2.288031e+02  2.143112e+02  2.135212e+02   

                 25            26            27            28            29  \
count  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06  2.918139e+06   
mean   7.271026e+02 -6.625340e+01  1.631537e+01 -8.637482e+01  1.517346e+02   
std    2.055724e+02  6.153334e+01  4.656628e+00  4.851313e+01  1.009000e+01   
min   -6.198046e+01 -3.585515e+02 -8.364693e+00 -3.140082e+02  1.074433e+02   
25%    8.137701e+02 -1.078521e+02  1.326106e+01 -1.186601e+02  

In [8]:
# Товары, для которых ищем похожие 'validation'
data_overview(validation)

0          1          2          3          4  \
Id                                                                    
100000-query -57.372734   3.597752 -13.213642 -125.92679  110.74594   
100001-query -53.758705  12.790300 -43.268543 -134.41762  114.44991   
100002-query -64.175095  -3.980927  -7.679249 -170.16093   96.44616   
100003-query -99.286860  16.123936   9.837166 -148.06044   83.69708   
100004-query -79.532920  -0.364173 -16.027431 -170.88495  165.45392   

                       5           6           7           8           9  \
Id                                                                         
100000-query  -81.279594 -461.003172  139.815720  112.880980   75.215750   
100001-query  -90.520130 -759.626065   63.995087  127.117905   53.128998   
100002-query  -62.377740 -759.626065   87.477554  131.270110  168.920320   
100003-query -133.729720   58.576403  -19.046660  115.042404   75.206730   
100004-query  -28.291668   33.931936   34.411217  128.903980  102.086914   

                     10          11         12         13         14  \
Id                                                                     
100000-query -131.89280 -140.968570 -57.987164 -22.868887  150.89552   
100001-query -153.71725  -63.951330 -52.369495 -33.390945  148.61950   
100002-query -220.30954  -31.378445  -8.788761   2.285323  133.26611   
100003-query -114.27196  -71.406456 -65.349320  24.377070   50.46730   
100004-query  -76.21417  -26.393860  34.423640  50.938890  157.68318   

                     15         16         17         18         19  \
Id                                                                    
100000-query   7.965574  17.622066 -34.868217 -216.13855 -80.908730   
100001-query -22.483830  15.164185 -56.202000 -153.61438 -79.831825   
100002-query -41.309080  14.305538 -18.231812 -205.53370 -78.160310   
100003-query -14.721335  15.069309 -46.682995 -176.60437 -78.690700   
100004-query -23.786497 -33.175415  -0.592607 -193.31854 -79.651030   

                      20           21         22          23         24  \
Id                                                                        
100000-query  -52.579520   263.363136  56.266876   66.924710  21.609911   
100001-query -101.055480  1203.537156  81.597130  101.018654  56.783424   
100002-query  -96.607670  1507.231274  -5.964200   34.937443 -56.086887   
100003-query -139.227450   325.547112   3.632292   74.929504  -4.802103   
100004-query  -91.889786  1358.481072  44.027733  121.527210  46.183000   

                      25          26         27         28         29  \
Id                                                                      
100000-query  813.770071  -32.782940  20.794031 -79.779076  156.30708   
100001-query   92.209628 -126.860340  10.382887 -38.523360  165.38391   
100002-query  813.770071  -13.200474  18.966661 -35.110190  151.36850   
100003-query  813.770071  -52.982597  15.644381 -54.087467  151.30914   
100004-query  433.623103  -82.233200  21.068508 -32.940117  149.26895   

                     30          31          32          33          34  \
Id                                                                        
100000-query -42.831330  -71.723335   83.283660 -304.174382    1.609402   
100001-query -77.840485 -169.538680  103.483240 -915.735701   16.109938   
100002-query -17.490252 -145.884300   15.533379 -655.395514   39.412827   
100003-query  21.088570 -134.507890   65.118960 -529.295053  131.565520   
100004-query   0.404718  -97.674530   81.719990 -825.644804    9.397169   

                     35         36         37          38         39  \
Id                                                                     
100000-query  55.834587 -29.474255 -139.16277 -126.038350 -62.643830   
100001-query  14.669937 -38.707085 -149.53838 -138.792920 -36.076176   
100002-query  62.554955   9.924992 -143.93462 -123.107796 -37.032475   
100003-query  67.642700 -22.884491 -145.90652  -86.917330 -11.86

Размер датасета: (100000, 72)
В датасете нет пропусков


0              1              2              3  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -85.302233       7.669724     -43.842474    -146.119797   
std        25.777321       4.956990      39.138775      20.495541   
min      -190.353330     -11.109877    -217.538420    -220.050890   
25%      -103.231790       4.292425     -69.680916    -160.210287   
50%       -85.296745       7.657888     -43.230835    -146.080365   
75%       -67.422104      10.993900     -17.765821    -132.119278   
max        14.427986      27.409784     134.859800     -57.381890   

                   4              5              6              7  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      111.635071     -73.273042    -441.396759      21.594242   
std        47.751576      28.515740     279.242585      66.478048   
min       -81.198990    -176.705030    -791.460243    -265.608430   
25%        79.101920     -92.789803    -740.620025     -21.804730   
50%       111.959330     -73.579283    -513.922978      22.138012   
75%       143.903735     -53.861139    -202.572440      66.097697   
max       302.065370      50.255325     109.632035     279.468900   

                   8              9             10             11  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      123.077725     126.001062    -141.215384     -69.340237   
std         6.504931      64.981697      41.568432      51.483403   
min        96.218350    -135.667300    -321.544250    -363.493320   
25%       118.657689      82.311205    -168.562772    -103.338022   
50%       123.062568     126.083595    -140.410225     -68.618967   
75%       127.538967     169.680500    -113.316906     -34.844935   
max       151.822560     392.827150      34.505280     170.344020   

                  12             13             14             15  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -42.626768       6.272022     149.006081     -23.961728   
std        40.858017      37.845939      34.031665      13.014670   
min      -204.299130    -144.100740     -14.816055     -78.597450   
25%       -70.147153     -19.089412     126.442184     -32.356516   
50%       -42.192298       6.420291     149.639465     -23.758115   
75%       -14.724666      31.930171     172.417195     -15.463551   
max       120.531970     177.762480     277.417850      36.471610   

                  16             17             18             19  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean       -2.009594     -13.151269    -138.540600     -79.162680   
std        20.993654      55.340411      48.790143       0.777431   
min       -86.843280    -272.565600    -329.860170     -82.665300   
25%       -16.392179     -49.942476    -171.695623     -79.665475   
50%        -2.075724     -12.401822    -139.000720     -79.139723   
75%        11.971967      24.229030    -106.607981     -78.636325   
max        92.661640     197.926830      82.683810     -76.177810   

                  20             21             22             23  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean      -97.356361    1301.395163      -3.439855      56.097825   
std        28.503558     371.543911      50.291410      35.599512   
min      -231.376650     136.846377    -208.900850     -84.224530   
25%      -115.390365    1248.126198     -36.891637      31.741733   
50%       -96.693777    1507.231274      -3.386905      55.461280   
75%       -78.378098    1507.231274      29.878309      79.815228   
max         9.297035    1557.293478     194.321810     203.892100   

                  24             25             26             27  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean       23.357262     741.156981     -64.986388      16.367853   
std        40.849607     193.199671      63.274623       4.845841   
min      -1

Можно заметить, что данные разделены на обучающую и тестовую выборки по порядку, ориентируясь на индексы

In [9]:
# Ответы к предыдущему набору товаров из 'validation' - 'answers'
data_overview(answers)

Expected
Id                        
100000-query  2676668-base
100001-query    91606-base
100002-query   472256-base
100003-query  3168654-base
100004-query    75484-base

Размер датасета: (100000, 1)
В датасете нет пропусков


Expected
count        100000
unique        91502
top     210304-base
freq              7

В данном разделе проекта мы:
- загрузили данные и вывели информацию о них
- определили, что в данных отсутствуют пропуски
- определили обучающие и валидационные выборки (набор признаков и таргет)  


## Подготовка данных к построению моделей

Обучающий датасет содержит набор признаков и таргет, которые нужно разделить

In [10]:
target = train['Target']
features = train.drop('Target', axis=1)

Переименуем признаки и таргет из валидационного сета

In [11]:
features_test = validation
target_test = answers['Expected']

Разделим обучающий датасет на тренировочную и валидационную выборки

In [12]:
# Сократим обучающую выборку до 50000 товаров
features_train, features_extra, target_train, target_extra = (
    train_test_split(features, target, test_size=0.9, random_state=12345)
)

Так как разброс в значениях данных большой, проведем масштабирование признаков

In [13]:
# Проведем масштабирование признаков

numeric = features_train.columns.to_list() # Запишем все количественные признаки в список numeric
# numeric

features_train_scaled = features_train.copy()
features_test_scaled = features_test.copy()
base_scaled = base.copy()

# Используем метод StandardScaler
# Он приводит все значения к стандартному нормальному распределению
scaler = StandardScaler()
scaler.fit(features_train_scaled[numeric])
features_train_scaled[numeric] = scaler.transform(features_train_scaled[numeric])
features_test_scaled[numeric] = scaler.transform(features_test_scaled[numeric])
base_scaled[numeric] = scaler.transform(base_scaled[numeric])

#features_train_scaled.head()
#features_valid_scaled.head()
#features_test_scaled.head()

In [16]:
# Удалим ненужные переменные
del (
    answers, base, base_url, delete_filepath, download_url, f,
    features, target, features_train, features_test, final_url,
    numeric, public_key, response, scaler, train, validation
)

В данном разделе мы:
- сохранили обучающую и тестовую выборки, отделив признаки от таргета
- провели масштабирование признаков  
  
Данные готовы к построению моделей

## Построение моделей

### FAISS

Построим модель FAISS с ускоренным поиском с помощью inverted lists. Посчитаем предсказания в два этапа.  
  
**Первый этап:**
- Сначала сгруппируем набор векторов из датасета `base` по частям с помощью k-means
- Для каждой части найдем центроид
- Далее будем искать для каждого вектора из `features_valid_scaled` наиболее близкий центроид
- В части с наиболее близким к вектору центроидом будем искать наиболее близкий вектор  
  
*Результатом первого этапа будет модель Faiss для предсказания 30 наиболее близких векторов к каждому из обучающей выборки*
  
**Второй этап:**
- Объединим полученыые предсказания и изначальные вектора, так у нас получится датасет с 72 * 2 = 144 признаками
- Добавим в признаки расстояние от каждого вектора из обучающей выборки до ближайших
- Посчитаем новый таргет, где 1 - предсказание модели соответствует рекомендации эксперта, 0 - нет
- Построим CatBoost, предсказывающий вероятность выбора конкретного товара в качестве наиболее похожего

#### Первый этап (FAISS)

Создадим индекс

In [17]:
n_cells = round(base_scaled.shape[0] ** 0.5) # количество кластеров
print(f'Разделим датасет на {n_cells} кластеров при создании индекса')

Разделим датасет на 1708 кластеров при создании индекса


Обучим индекс и добавим в него вектора из `base_scaled` для получения предсказаний

In [18]:
dim = features_train_scaled.shape[1] # размерность вектора признака
quantiser = faiss.IndexFlatL2(dim)
index = faiss.IndexIVFFlat(quantiser, dim, n_cells) # создание индекса

In [19]:
index.train(np.ascontiguousarray(features_train_scaled.values).astype('float32')) # обучим индекс на трейне
print(index.is_trained) # проверим обучение

index.add(np.ascontiguousarray(base_scaled.values).astype('float32')) # добавим вектора в индекс

True


In [20]:
# Пронумеруем id товаров, запишем в словарь
base_index = {n: v for n, v in enumerate(base_scaled.index.to_list())}
# base_index

Посчитаем предсказания на обучающей выборке и оценим качество модели с гиперпараметром nprobe = 300. Так как нам нужно будет использовать код повторно для оценки качества модели на тестовой выборке, напишем функцию для вычисления предсказаний

In [21]:
'''
Функция принимает на вход заскейленные признаки и таргет
Функция возвращает просчитанные индексы ближайших векторов и расстояния до них, а также метрику accuracy@30 и время подсчета
'''

def faiss_target_formation(features_faiss, target_faiss):

    start = time.time()
    index.nprobe = 300 # обозначим количество кластеров для поиска соседей
    vecs_new, idx_new = index.search(np.ascontiguousarray(features_faiss.values).astype('float32'), 30) # найдем вектора на трейне
    end = time.time()

    acc = 0
    for t, el in zip(target_faiss.values.tolist(), idx_new.tolist()): # посчитаем accuracy@30 для модели с nprobe=n
        acc += int(t in [base_index[r] for r in el])

    acc_faiss = 100 * acc / len(idx_new)

    faiss_time = end - start

    return vecs_new, idx_new, faiss_time, acc_faiss

In [ ]:
'''
vecs, idx, faiss_time, acc_faiss = faiss_target_formation(features_train_scaled, target_train)
# print(acc_faiss, faiss_time)
'''

In [22]:
acc_faiss = 74.18
faiss_time = 507.5104320049286

print(f'Точность алгоритма faiss на nprobe=100 и 1708 кластерах: {acc_faiss}\n'
    f'Время подбора ближайших векторов для трейна модели faiss: {faiss_time}')

Точность алгоритма faiss на nprobe=100 и 1708 кластерах: 74.18
Время подбора ближайших векторов для трейна модели faiss: 507.5104320049286


In [ ]:
'''
# Скачаем на диск полученные индексы для train
with open('/content/drive/My Drive/idx.txt', 'w') as f:
    np.savetxt(f, idx)

# Скачаем на диск расстояния до ближайших векторов для каждого товара для train
with open('/content/drive/My Drive/vecs.txt', 'w') as f:
    np.savetxt(f, vecs)
'''

In [23]:
with open('/content/drive/My Drive/idx.txt', 'r') as f:
    idx = np.loadtxt(f)

with open('/content/drive/My Drive/vecs.txt', 'r') as f:
    vecs = np.loadtxt(f)

Метрика accuracy@n получилась 74.18 при nprobe=100. Попробуем улучшить результат на втором шаге, используя CatBoost

#### Второй этап (CatBoost)

Напишем функцию для формирования признаков и таргета для CatBoost

In [24]:
'''
Функция принимает на вход признаки и таргет faiss
Функция возвращает склеенные признаки старых векторов и ближайших к ним, таргет для катбуста (1-вектор правильно побобран моделью, 0-неправильно)
'''

def catboost_features_formation(features_faiss, target_faiss, idx_faiss, vecs_faiss):

    rows = np.repeat(features_faiss.values, 30, axis=0).shape[0]
    cols = np.repeat(features_faiss.values, 30, axis=0).shape[1]

    added_features = np.zeros(shape=(rows, cols)) # Создадим датасет размерностью 100,000 * 50 на 72 для новых признаков
    idx_reshaped = idx.reshape(rows) # запишем индекс в более удобной форме
    vecs_faiss = vecs_faiss.reshape(-1, 1)

    for row in range(rows):
         added_features[row] = base_scaled.loc[base_index[idx_reshaped[row]]]

    features_catboost = np.concatenate(
        (np.repeat(features_faiss.values, 30, axis=0), added_features),
            axis=1
    )

    features_catboost = np.concatenate(
        (features_catboost, vecs_faiss),
            axis=1
    )

    target_catboost = []
    for t, el in zip(target_faiss.values.tolist(), idx_faiss.tolist()):
        for r in el:
             if base_index[r] == t:
                 target_catboost.append(1)
             else:
                 target_catboost.append(0)

    target_catboost = np.array(target_catboost)

    # target_catboost.sum() == acc

    return features_catboost, target_catboost

In [25]:
features_catboost, target_catboost = catboost_features_formation(features_train_scaled, target_train, idx, vecs)
# 100 * target_catboost.sum() / len(idx) == acc_faiss

Подберем гиперпараметры для CatBoost (второго этапа метчинга)

In [ ]:
'''
start = time.time()
grid_space_catboost = {
    'iterations': [550],
    'learning_rate': [0.01, 0.1, 0.2],
    'depth': np.arange(5, 8)
}

model_catboost = CatBoostClassifier(random_state=12345, verbose=False,
                                    auto_class_weights='Balanced',
                                    eval_metric='Accuracy')

grid_search_result = model_catboost.grid_search(grid_space_catboost, X=features_catboost,
                                                y=target_catboost, cv=3, verbose=False)
end = time.time()
'''


bestTest = 0.8489783634
bestIteration = 549


bestTest = 0.9057586594
bestIteration = 521


bestTest = 0.9055535656
bestIteration = 290


bestTest = 0.8580301076
bestIteration = 548


bestTest = 0.8998896501
bestIteration = 444


bestTest = 0.9028305796
bestIteration = 175


bestTest = 0.8685708152
bestIteration = 548


bestTest = 0.8984212011
bestIteration = 251


bestTest = 0.8994442512
bestIteration = 138

Training on fold [0/3]

bestTest = 0.9008142136
bestIteration = 329

Training on fold [1/3]

bestTest = 0.910775199
bestIteration = 528

Training on fold [2/3]

bestTest = 0.9005224652
bestIteration = 364



In [26]:
# best_params = grid_search_result['params']
best_params = {'depth': 5, 'iterations': 550, 'learning_rate': 0.1}
print(f'Лучшие гиперпараметры модели CatBoost: {best_params}')

# catboost_time = end - start
catboost_time = 1168.5561311244965
print(f'Впемя обучения модели CatBoost: {catboost_time}')

Лучшие гиперпараметры модели CatBoost: {'depth': 5, 'iterations': 550, 'learning_rate': 0.1}
Впемя обучения модели CatBoost: 1168.5561311244965


In [27]:
model_catboost = CatBoostClassifier(random_state=12345, verbose=False,
                                    depth=5, iterations=550,
                                    learning_rate=0.1, auto_class_weights='Balanced',
                                    eval_metric='Accuracy')

model_catboost.fit(features_catboost, target_catboost)

predictions_train = model_catboost.predict_proba(features_catboost)[:, 1]

Основываясь на accuracy score на кросс-валидации, качество модели улучшилось после добавления CatBoost. Посчитаем accuracy@5 на обучающей выборке, чтобы в этом убедиться. Напишем функцию для ее подсчета

In [28]:
def accuracy_5(predictions, target):

    # Выведем индексы номеров, на которых стоят 5 наиболее вероятных товаров для каждого из обучающей выорки
    predictions = pd.DataFrame(predictions.reshape(-1, 30))
    listed_idx = []
    for row in range(predictions.shape[0]):
        listed_idx.append(predictions.loc[row, :].sort_values(ascending=False)[:5].index.to_list())


    # Поставим 1 на место пяти наиболее вероятных товаров, предсказанных катбустом
    predictions = np.zeros(shape=predictions.shape)
    for row in range(predictions.shape[0]):
        for elem in listed_idx[row]:
            predictions[row][elem] = 1

    # Посчитаем accuracy@5
    target = target.reshape(-1,)
    predictions = predictions.reshape(-1,)
    acc = len([i for i, j in zip(target, predictions) if i == j and i == 1]) / target.reshape(-1, 30).shape[0]

    return acc

In [29]:
accuracy_5(predictions_train, target_catboost) * 100

73.65

Accuracy@5 после катбуста немного упала. Нужно посчитать метрику на тестовой выборке, чтобы убедиться, что второй этап с градиентным бустингом нужен

In [30]:
del idx, vecs, features_catboost, target_catboost

## Проверка модели на тестовой выборке

In [31]:
# Сократим тестовую выборку до 50000 товаров
features_test, features_extra_test, target_test, target_test = (
    train_test_split(features_test_scaled, target_test, test_size=0.9, random_state=12345)
)

In [46]:
'''
# Посчитаем ближайшие векторы и расстояния для сокращенной тестовой выборки
vecs_test, idx_test, faiss_time, acc_faiss = faiss_target_formation(features_test, target_test)
'''

In [26]:
'''
# Скачаем на диск полученные индексы для train
with open('/content/drive/My Drive/vecs_test.txt', 'w') as f:
    np.savetxt(f, vecs_test)

# Скачаем на диск расстояния до ближайших векторов для каждого товара для train
with open('/content/drive/My Drive/idx_test.txt', 'w') as f:
    np.savetxt(f, idx_test)
'''

In [32]:
with open('/content/drive/My Drive/idx_test.txt', 'r') as f:
    idx = np.loadtxt(f)

with open('/content/drive/My Drive/vecs_test.txt', 'r') as f:
    vecs = np.loadtxt(f)

In [35]:
acc = 0
for t, el in zip(target_test.values.tolist(), idx.tolist()): # посчитаем accuracy@30 для модели с nprobe=n
    acc += int(t in [base_index[r] for r in el])

acc_faiss = 100 * acc / len(idx)
acc_faiss

0.02

Метрика на faiss получилась очень низкая

In [36]:
# Обозначим признаки и таргет для катбуста, найдем предсказания из катбуста
features_catboost_test, target_catboost_test = catboost_features_formation(features_test, target_test, idx, vecs)
predictions_test = model_catboost.predict_proba(features_catboost_test)[:, 1]

In [37]:
accuracy_5(predictions_test, target_catboost_test)

0.0